modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableViewSkinBase.java

```diff
@@ -638,10 +692,17 @@ Region getColumnReorderLine() {
     }
 
     /**
-     * Function used to scroll the container down by one 'page', although
-     * if this is a horizontal container, then the scrolling will be to the right.
+     * Scrolls the container down by one 'page'. The direction is determined by the orientation ({@link
+     * VirtualFlow#isVertical()}) of the {@code VirtualFlow} (down for vertical, right for horizontal).
+     * <p>
+     * If the last visible cell is selected (or f{@code isFocusDriven} is {@code true}), it will be made visible at the
+     * very top.
+     *
```

Comment:
Then ask Kevin or Ajit for an explanation.

If your assumption is true, then something like:
"Scrolls the container [maybe "content" is better?] down so that the last cell becomes the first cell" should be good.

Reply:
Ok I've read this "Move the selection down on the last fully visible item on the current page. If the last fully visible item is selected, move the selection on the last fully visible item on the next page." from https://wiki.openjdk.java.net/display/OpenJFX/TableView+User+Experience+Documentation

And that concurs what I've deduced from the code. This seems quite odds but the specification are what they are. 
So "Scrolls the content down so that the last cell becomes the first cell. If the last visible cell is selected, scrolls the content down so that the last cell on the next page because visible"? That's rather odd because we're using the definition of a page inside the definition of a page...

Reply:
I don't see the definition of a page there, but it really seems like a "page" is the cell/row content of the viewport. Can you test empirically with a `TableView` how it behaves?

Reply:
Allright I've been able to test on a TableView. When pressing "page down", the selection is set on the last fully visible cell/row. So a page is definitely the content of the viewport. Then if I press again "page down", it will scrolll so that the last visible cell that was selected is displayed on top.
In the description, I would empathized the fact that this method is called when the selection is changing.
Maybe something like that :
"Returns the index to select after a page scroll.
The last fully visible cell index on the current viewport is always returned. If this index is already selected, a scroll is happening first, so that it is displayed at the top of the viewport. Then the last fully visible cell index of the viewport is returned.
If {@code isFocusDriven} is set to {@code true}, the focused cell is considered instead of the selected cell."

Reply:
Alright, so that's confirmed.

I would rephrase starting from the 2nd sentence to:
"The returned index belongs to the last fully visible cell in the viewport, after scrolling (if needed). Scrolling is done only if that cell is already selected, in which case that cell is made visible at the top of the viewport.
If `{@code isFocusDriven}` is ["set to" is optional] `{@code true}`, the focused cell is used instead of the selected cell."

Does the last sentence apply only to the scrolling, or to the returned result also? That is, if it's `true`, is the first sentence effectively "Returns the index to focus after a page scroll."?

Reply:
I just checked in the code and it's the case. If `isFocusDriven` is `true`, the index will be focused after a page scroll, not selected.

Reply:
This method is starting to be cumbersome because it can either focus or select, and it can either scroll the content or not (even though it's ca;;ed "onScroll..."). It seems to me like it could be 2 methods and renamed to "selectOnPageUp" and "focusOnPageUp".

In any case, maybe:

> Returns the index of the selected (or focused, if `{@code isFocusDriven}` is `{@true}`) cell after a page scroll operation [added "operation" because actual scrolling doesn't always happen]. If the selected/focused cell is not the last fully visible cell, then the last fully visible cell is selected/focused. Otherwise, the content is scrolled such that the cell is made visible at the top of the viewport (and the new last fully visible cell is selected/focused instead).

I'm afraid to ask what happens if the selected/focused cell is the "real" last cell - the one not fully visible. Does it not behave like the last fully visible cell?

Reply:
The only good thing about this behavior is that the docs are still accurate...

Honestly, this behavior sounds like a bug to me.

Reply:
I totally agree with you but it seems like changing some "package-protected" methods to "protected" is a huge deal. I can't even imagine what it will be if I suggest changing the internal code and behavior of the method to be honest.
I think we should document it like we did, and maybe open a bug/improvement ticket that can be resolved later or in a different PR.

Reply:
to put coal into the fire: there's an [open issue concerning scrolling behavior](https://bugs.openjdk.java.net/browse/JDK-8197536) on list/table ;)